### ISSUE TITLE
Clang 22 now warns on `__COUNTER__`. It shouldn't

### ISSUE THREAD

**Issue by brevzin** [llvm/llvm-project#196557]
Simple example:

```cpp
int x = __COUNTER__;
```

Compiling with `-pedantic`, this compiles cleanly on every version of clang. Up until 22.1.0, where I get:

```
<source>:1:9: warning: '__COUNTER__' is a C2y extension [-Wc2y-extensions]
    1 | int x = __COUNTER__;
      |         ^
```

It might be a C2y extension now, but it was also a clang (and gcc and edg and msvc) extension before. The fact that it's _now_ standard shouldn't be a reason to start warning on it all of a sudden. It's been so ubiquitously supported for so long that many people assume that it has always been standard — probably why clang didn't even warn `-pedantic` before. 

It seems to me that the only plausible action for me to take in response to this warning is to disable this warning. So probably it shouldn't warn. Plus, disabling this warning might hide other C2y extension warnings that I might actually want to take action against. I don't know enough about C2y to know what those other extensions might be, but now that I've disabled this warning, I guess I'll never find out.

**Comment by AaronBallman** [comment]
This is by design; when compiling with `-pedantic`, we emit diagnostics for extensions; that we missed the diagnostic previously was unintentional. That said, we did adjust the behavior in 22.x around `-pedantic-errors` and `-Wno-*` flag handling: https://github.com/llvm/llvm-project/blob/c0262e742787e5997a6df2b4b56830df00a6ff11/clang/docs/ReleaseNotes.rst?plain=1#L581

(I'm assuming you're talking about cases where `__COUNTER__` really is an extension and not in `-std=c2y` mode or something. If the diagnostic is triggering in cases where `__COUNTER__` is not an extension, that would be really good to know because that's almost certainly a bug.)

**Comment by AaronBallman** [comment]
This said, I think I would be in support of some diagnostic flag to control `__COUNTER__` extension diagnostics specifically so users can do `-pedantic -Wno-complaining-about-counter` if they want to know about extensions except for `__COUNTER__`... but I also feel like that opens the question of whether we should do that consistently for all extensions and how to do that in a way that scales. But this one is sufficiently disruptive I think it may make sense even if we don't do it consistently.

**Comment by AaronBallman** [comment]
> Hi, I’d like to work on this issue if it’s still open.

Feel free; I think the direction we're thinking of going is described in: https://github.com/llvm/llvm-project/issues/196557#issuecomment-4408482256 though we've not nailed down what to name the flag. Maybe something like `-Wcounter-extension`? There's also a PR currently opened which will silence this diagnostic when expanded from system macros, so we should test how the new flag interacts (https://github.com/llvm/llvm-project/pull/196689).

**Comment by brevzin** [comment]
Why start warning on `__COUNTER__` at all though? Clang hasn't ever warned on this before, and it's finally even standardized. Are there users that actually want a warning here? Why?

**Comment by AaronBallman** [comment]
> Why start warning on `__COUNTER__` at all though?

It's an extension and we have a diagnostic flag for "tell me about all extensions".

> Clang hasn't ever warned on this before, and it's finally even standardized. 

Clang's policy is to add these missing extension warnings when we're updating the code. So this got a warning *because* it was standardized; that's when we noticed it was unintentionally missing.

> Are there users that actually want a warning here? Why?

Yes; for the same reason anyone turns on `-pedantic`: to be told about use of extensions. `__COUNTER__` is well-supported by existing C and C++ implementations, but it's not ubiquitous (https://godbolt.org/z/qjcdaa7s1, https://godbolt.org/z/cE4sWfbvz) and because it's not standard, it's not guaranteed to be supported by compilers in the future either.

**Comment by ojhunt** [comment]
> Why start warning on `__COUNTER__` at all though? Clang hasn't ever warned on this before, and it's finally even standardized. Are there users that actually want a warning here? Why?

This is what I've argued, we're only introducing the warning at the point it is being standardized, which seems counter to commonsense even if this is an extension, however it would appear this ship has sailed.

Certainly is a chopping block warning from my PoV. It depends on what things look like with #196689 - as far as I'm concerned if this is still causing build failures (via -Werror) with the system header fix it's a no-go.

I'm pondering extended the nowerror flag on warning diagnostics to cover extension warnings as well which might also work, but if it remains noisy I'll simply be removing the warning downstream, even though that's obnoxious for its own reasons.

**Comment by ojhunt** [comment]
Ok, I did just check, `__COUNTER__` is supported by: clang, gcc, icc, msvc, icx (???), and _apparently_ arms older compilers. You can find some niche compilers that don't support it, but they also don't support an array of standard features either. They're no choosing to not support `__COUNTER__`: they don't have the time or resources to support C99 either, so using them as a "proof of `__COUNTER__` not being supported everywhere" is somewhat disingenuous.

Moreover: even GCC does not warn about `__COUNTER__`: not with -Wall, not with -pedantic, not with -Wpedantic.

Using "it is now becoming formally standardized" as a reason to then make a breaking (for -Werror builds) is not reasonable: neither logically, nor from a basic compatibility point of view.

**Comment by AaronBallman** [comment]
> Ok, I did just check, `__COUNTER__` is supported by: clang, gcc, icc, msvc, icx (???),

icx is Clang-based, so it does support `__COUNTER__`. (FWIW, icc is EDG-based)

> and _apparently_ arms older compilers. You can find some niche compilers that don't support it, but they also don't support an array of standard features either. They're no choosing to not support `__COUNTER__`: they don't have the time or resources to support C99 either, so using them as a "proof of `__COUNTER__` not being supported everywhere" is somewhat disingenuous.

It's proof that it's not a fully portable construct (which is something others have claimed as a reason to not issue extension warnings for it). And whether the compiler is niche or not doesn't matter, there are multiple compilers which don't support the extension. From my quick but slightly more thorough testing: EDG (in default mode), LC3, Norcroft ARM, Tendra, and ORCA/C don't support it (at least out of the box). Not supporting C99 isn't really relevant here; the diagnostic is about portability, so what's more important is whether there are implementations which don't support the feature and there are plenty of examples of that. (I've not started looking at the embedded space but I suspect there's also implementations there which don't support the extension.)

> Moreover: even GCC does not warn about `__COUNTER__`: not with -Wall, not with -pedantic, not with -Wpedantic.

I have not found evidence that this was a result of an intentional decision; did you see something to the contrary? Clang often diagnoses things GCC does not and vice versa, so barring a discussion, I don't know how much weight to give "but GCC doesn't do it", particularly for an extension this old.

> Using "it is now becoming formally standardized" as a reason to then make a breaking (for -Werror builds) is not reasonable: neither logically, nor from a basic compatibility point of view.

Adding new (correctly issued) diagnostics is not a breaking change; this was a false negative that was fixed and we've demonstrated why the construct is not portable and thus extension diagnostics are useful to some users.

**Comment by fwyzard** [comment]
> extension diagnostics are useful to some users

Has _any_ user actually come forward saying that this is useful for them?

**Comment by shafik** [comment]
> > extension diagnostics are useful to some users
> 
> Has _any_ user actually come forward saying that this is useful for them?

What our users find useful is quite diverse, I am continually surprised by what I hear when having in person conversations with various developers. Some of it is particular to their organization, some of it is personal preference based on what caused them pain before, some of it could be regulatory etc

**Comment by brevzin** [comment]
> > > extension diagnostics are useful to some users
> > 
> > 
> > Has _any_ user actually come forward saying that this is useful for them?
> 
> What our users find useful is quite diverse, I am continually surprised by what I hear when having in person conversations with various developers. Some of it is particular to their organization, some of it is personal preference based on what caused them pain before, some of it could be regulatory etc

So... no users then? `__COUNTER__` has been around for a very long time while being an extension, so if somebody wanted a warning for it, I'd expect somebody to have asked about it. But I can't find anybody having done so, and you're not offering either. 

In the meantime, this warning does cause problems (as @cor3ntin even [warned about](https://github.com/llvm/llvm-project/pull/162662#issuecomment-3389994326) in the PR). Adding this warning led to a bunch of issues I've found in other repos

* breaking [google benchmark](https://github.com/google/benchmark/issues/2057) (they reacted by silencing the C2y extensions warning)
* breaking [doctest](https://github.com/doctest/doctest/issues/1042) (they reacted by switching to `__LINE__` specifically for Clang 22+)
* breaking [Catch2](https://github.com/catchorg/Catch2/issues/3076) (they reacted by silencing the C2y extensions warning)
* breaking us, leading me to open this issue (we reacting by silencing the C2y extensions warning)

In all of these cases, this is extra work that people had to do to work around this unhelpful warning.

**Comment by ojhunt** [comment]
Oh yeah, __COUNTER__ took out the OS build - and we needed to add the fix for system header macros and a reasonably sizable number of per-project overrides. The granularity meant the per-project settings suppressed all c2y extension warnings.

**Comment by erichkeane** [comment]
What is being missed here is that this SHOULD have always warned.  It was a bug that it didn't.  Clang is always supposed to diagnose extensions, no matter where they come from.  In this case, the fact that it got standardized is what made us notice that.  We typically diagnose it as the most 'authoritative' version (that is, since it was standardized the message is C2y, previously it should have been GNU, or some vague "this is an extension from the eather").

That said, arguments of "this broke X!" with a new warning are bupkis.  `-Werror` means, "Any time the compiler engineers add a new warning to a group I've opted into, please break me".  The fact that it DID break things is evidence that this is a good warning in fact, since that is exactly what those projects asked for!

Finally, as this is so heavily used, I think it definitely deserves a 'sub-warning-group' so it can be disabled separately.  We were on our way there in the original report with CI, but seems like we got distracted by other things.  But that is how I would see this bug being 'fixed'.

**Comment by timsong-cpp** [comment]
> `-Werror` means, "Any time the compiler engineers add a new warning to a group I've opted into, please break me". The fact that it DID break things is evidence that this is a good warning in fact, since that is exactly what those projects asked for!

That argument would equally hold for a warning that fires on tokens more than 3 characters in length. Is that a good warning too?

**Comment by erichkeane** [comment]
> > `-Werror` means, "Any time the compiler engineers add a new warning to a group I've opted into, please break me". The fact that it DID break things is evidence that this is a good warning in fact, since that is exactly what those projects asked for!
> 
> That argument would equally hold for a warning that fires on tokens more than 3 characters in length. Is that a good warning too?

No?  And asking that question requires a vast-misreading of what I said.

**Comment by timsong-cpp** [comment]
> The fact that it DID break things is evidence that this is a good warning in fact

That's what you said. Did I misunderstand it?

**Comment by erichkeane** [comment]
> > The fact that it DID break things is evidence that this is a good warning in fact
> 
> That's what you said. Did I misunderstand it?

Seemingly so?  "There is evidence that..." does not mean "It is proven that...".  You're taking `A + B + C + D => E`, and extrapolating that: `C => E` here.

**Comment by ojhunt** [comment]
The core problem here is that `__COUNTER__` has essentially existed forever, and the warning was only introduced at the point of standardization which remains absurd to me.

As mentioned when I filed the original "this breaks a large number of builds", this broke a large number of projects for no reason whatsoever.

But also for cases like this the only project solutions is to manually annotate every use and every macro to disable the warning, or to disable the extension warning globally. e.g The addition of `__COUNTER__` to the "this now issues an extension warning" has the functional effect of having projects lose the warning even for extensions that do matter.

If we absolutely _must_ issue a diagnostic for this (which I still disagree with) we need a separate diagnostic group for "widely used features that we are now going to flag as an extension" that has the flag so that they don't become errors under `-Werror`.

Again, the decision to add an extension warning to the feature at the point it was actually standardized remains fundamentally absurd to me.

**Comment by erichkeane** [comment]
> The core problem here is that `__COUNTER__` has essentially existed forever, and the warning was only introduced at the point of standardization which remains absurd to me.
> 
> As mentioned when I filed the original "this breaks a large number of builds", this broke a large number of projects for no reason whatsoever.
> 
> But also for cases like this the only project solutions is to manually annotate every use and every macro to disable the warning, or to disable the extension warning globally. e.g The addition of `__COUNTER__` to the "this now issues an extension warning" has the functional effect of having projects lose the warning even for extensions that do matter.
> 
> If we absolutely _must_ issue a diagnostic for this (which I still disagree with) we need a separate diagnostic group for "widely used features that we are now going to flag as an extension" that has the flag so that they don't become errors under `-Werror`.
> 
> Again, the decision to add an extension warning to the feature at the point it was actually standardized remains fundamentally absurd to me.

The point of `-pedantic` is to be pedantic.  If "warn you about something you don't care about, but is TECHNICALLY correct" doesn't fit the definition of pedantic, I don't know the meaning of the word :)  

I agree that it stinks we added this so late, as this diagnostic should have been in the compiler decades ago.  Unfortunately we cannot time-travel, so we can't go back in time to fix it decades ago (IF I could time-travel, I'd probably rename `-pedantic` to `-Wwell-ACKTSHUALLY`).  The 'standardization' is frankly just how we discovered it, and is pretty irrelevant here (other than giving us... something... to put in the message).

> we need a separate diagnostic group 
yeah, I believe @cor3ntin has promised us he's going to make a special -W flag for this one, and I think basically everyone here agrees to that.

.

**Comment by jwakely** [comment]
> I agree that it stinks we added this so late, as this diagnostic should have been in the compiler decades ago. Unfortunately we cannot time-travel, so we can't go back in time to fix it decades ago

Correcting the mistake now is not the only option, saying  "oops, this should have been a prefabricated lpedantic warning, but it's too late/disruptive to do that now" is also an option.

**Comment by AaronBallman** [comment]
This feature was implemented very early on in Clang's lifetime, before we had policies around diagnostics and quality of implementation, etc. Like many such features from that era, we're missing diagnostics that are expected today. However, these diagnostics are not the highest priority work and so we typically add them only when we're updating the feature for some reason. In this case, the feature got standardized and that is why we updated the diagnostic behavior. But to be clear: this is behaving by design and is not a bug, it was an intentional correction to our behavior. This diagnostic does not break code. The user has to opt in to asking for the diagnostic and new (correctly implemented) diagnostics are never considered a breaking change. It's also not novel to produce extension diagnostics for preprocessor features (https://godbolt.org/z/rnMcvcE66).

That said, I should have predicted the fallout here given how old the extension is and where it sits in the compiler stack (mea culpa!). I should have introduced this with its own diagnostic group so it could be controlled separately from `-pedantic`. I've posted a PR to do that and am going to try to get it backported to the 23 release. This is something for us to keep in mind for the next time we go through this.
